File: code/ch19/19.4.5. 668.findKthNumber.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
  int countNotGreater(int mid, int m, int n) {
    int cnt = 0;
    for (int i = 1; i <= m; i++)
      cnt += min(mid / i, n);
    return cnt;
  }
  int findKthNumber(int m, int n, int k) {
    int lo = 1, hi = m * n;
    while (lo < hi) {
      int mid = (lo + hi) / 2;
      if (countNotGreater(mid, m, n) < k)
        lo = mid + 1;
      else
        hi = mid;
    }
    return lo;
  }
};

File: code/ch19/19.4.5. 668.findKthNumber.py
class Solution:
    def findKthNumber(self, m: int, n: int, k: int) -> int:
        lo = 1
        hi = m * n

        def countNotGreater(mid: int, m: int, n: int) -> int:
            cnt = 0
            for i in range(1, m + 1):
                cnt += min(mid // i, n)
            return cnt

        while lo < hi:
            mid = (lo + hi) // 2
            if countNotGreater(mid, m, n) < k:
                lo = mid + 1
            else:
                hi = mid
        return lo
